qiskit/extensions/hamiltonian_gate.py

```diff
@@ -0,0 +1,151 @@
+        if input_dim != output_dim or 2**num_qubits != input_dim:
+            raise ExtensionError(
+                "Input matrix is not an N-qubit operator.")
+
+        self._qasm_name = None
+        self._qasm_definition = None
+        self._qasm_def_written = False
+        # Store instruction params
+        super().__init__('hamiltonian', num_qubits, [data, time], label=label)
+
+    def __eq__(self, other):
+        if not isinstance(other, HamiltonianGate):
+            return False
+        if self.label != other.label:
+            return False
+        # Should we match unitaries as equal if they are equal
+        # up to global phase?
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=True)
+        times_eq = self.params[1] == other.params[1]
+        return operators_eq and times_eq
+
+    def to_matrix(self):
+        """Return matrix for the unitary."""
+        unbound_t = False
+        if isinstance(self.params[1], ParameterExpression):
+            unbound_t = len(self.params[1].parameters) > 0
+        if isinstance(self.params[1], Number) or not unbound_t:
+            # pylint: disable=no-member
+            return scipy.linalg.expm(1j * self.params[0] * self.params[1])
+        else:
+            raise NotImplementedError("Unable to generate Unitary matrix for "
+                                      "unbound t parameter {}".format(self.params[1]))
+
+    def inverse(self):
+        """Return the adjoint of the unitary."""
+        return self.adjoint()
+
+    def conjugate(self):
+        """Return the conjugate of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), -self.params[1])
+
+    def adjoint(self):
+        """Return the adjoint of the unitary."""
+        return HamiltonianGate(self.params[0], -self.params[1])
+
+    def transpose(self):
+        """Return the transpose of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), self.params[1])
+
+    def _define(self):
+        """Calculate a subcircuit that implements this unitary."""
+        return UnitaryGate(self.to_matrix())
+
+    def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
+        r"""Return controlled version of gate, relying on control implementation in Unitary.
+
+        Args:
+            num_ctrl_qubits (int): number of controls to add to gate (default=1)
+            label (str): optional gate label
+            ctrl_state (int or str or None): The control state in decimal or as a
+                bit string (e.g. '1011'). If None, use 2**num_ctrl_qubits-1.
+
+        Returns:
+            UnitaryGate: controlled version of gate.
+
+        Raises:
+            QiskitError: invalid ctrl_state
+        """
+        return self._define().control(num_ctrl_qubits=num_ctrl_qubits,
+                                      label=label,
+                                      ctrl_state=ctrl_state)
+
```

Comment:
If you implement `_define` as is done with the other gates you should be able to use `control` from the base class 🙂